nest_elephant_tvb/translation/mpi_translator.py

```diff
@@ -3,192 +3,237 @@
+    NOTE: First refactored version -> not pretty, not final. 
+    
+    TODO: Ugly: 'store' and 'analyse' objects passed through all the way from the beginning.
+    TODO: Discuss communication protocol of NEST<->transformer and transformer<->TVB
     '''
+
     count=0
     status_ = MPI.Status()
     while True: # FAT END POINT
-        # wait until the translator accept the connections
+        # TODO: this communication has the 'rank 0' problem described in the beginning
         accept = False
         logger.info("Nest to TVB : wait to send " )
         while not accept:
-            req = comm.irecv(source=MPI.ANY_SOURCE,tag=MPI.ANY_TAG)
+            req = comm_sender.irecv(source=MPI.ANY_SOURCE,tag=MPI.ANY_TAG)
             accept = req.wait(status_)
         logger.info(" Nest to TVB : send data status : " +str(status_.Get_tag()))
         if status_.Get_tag() == 0:
-            # send the rate when there ready
-            while status_data[0] != 0: # FAT END POINT
+            # wait until the receiver has cleared the buffer, i.e. filled with new data
+            while databuffer[-1] != 0:
                 time.sleep(0.001)
                 pass
-            times,data=analyse.analyse(count,buffer[0])
-            with lock_status:
-                if status_data[0] != 2:
-                    status_data[0] = 1
+            # TODO: All science/analysis here. Move to a proper place.
+            times,data = _analyse(count, databuffer, store, analyse)
+            
+            # Mark as 'ready to receive next simulation step'
+            databuffer[-1] = 1
+            
+            ############ OLD Code
+            # TODO: this communication has the 'rank 0' problem described in the beginning
             logger.info("Nest to TVB : send data :"+str(np.sum(data)) )
             # time of stating and ending step
-            comm.Send([times, MPI.DOUBLE], dest=status_.Get_source(), tag=0)
+            comm_sender.Send([times, MPI.DOUBLE], dest=status_.Get_source(), tag=0)
             # send the size of the rate
             size = np.array(int(data.shape[0]),dtype='i')
-            comm.Send([size,MPI.INT], dest=status_.Get_source(), tag=0)
+            comm_sender.Send([size,MPI.INT], dest=status_.Get_source(), tag=0)
             # send the rates
-            comm.Send([data,MPI.DOUBLE], dest=status_.Get_source(), tag=0)
+            comm_sender.Send([data,MPI.DOUBLE], dest=status_.Get_source(), tag=0)
+            ############ OLD Code end
         elif status_.Get_tag() == 1:
             # disconnect when everything is ending
-            with lock_status:
-                status_data[0] = 2
             break
         else:
             raise Exception("bad mpi tag"+str(status_.Get_tag()))
         count+=1
-    logger.info('communication disconnect')
-    comm.Disconnect()
-    logger.info('end thread')
-    return
+    logger.info('NEST_to_TVB: End of send function')
+
+
+def _analyse(count, databuffer, store, analyse):
```

Comment:
you can move this function in science_nest_to_TVB. If you have time, you can redesign by merging store and analyse in one function.

Reply:
Is on the roadmap.